[nxp_s32_gmac][tja1103] Handle link status change from net shell and link partner

sumitbatra-nxp opened the pull request:
These changes enable nxp s32 gmac ethernet driver to gracefully handle
net shell commands - net iface down and up and 
link connection or disconnection between tja1103 ethernet phy and its link partner.

PetervdPerk-NXP commented:
Why not implement the TJA1103 phy interrupt status? That was the reason I didn't implement the link changes because I didn't want to spend a kernel worker to actively poll the phy status.

sumitbatra-nxp commented:
> Why not implement the TJA1103 phy interrupt status? That was the reason I didn't implement the link changes because I didn't want to spend a kernel worker to actively poll the phy status.

@PetervdPerk-NXP .. Thanks for pointing out.. I missed the interrupt line connection to the SoC in the CANHUBK3 schematics.. But lets keep both interrupt and polling in case someone doesn't have an interrupt line

manuargue reviewed (changes requested):
Pls split the changes in two commits for clarity. One for introducing the link state polling in the TJA1103 driver, and another for integrating a PHY driver into the S32 GMAC driver.

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -78,6 +80,24 @@ static inline struct net_if *get_iface(struct eth_nxp_s32_data *ctx, uint16_t vl
+
+	if (state->is_up) {
+		LOG_DBG("Link up");
+		net_eth_carrier_on(ctx->iface);
```
I knew this would come up, but just wanted to get through with CANHUBK3 use case first.

sumitbatra-nxp replied:
This is work in progress.. and just for my ease I'll split the commits between MAC and PHY once autoneg (speed and duplex) change is made

sumitbatra-nxp replied:
done !!

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -181,12 +204,15 @@ static int eth_nxp_s32_start(const struct device *dev)
 	Gmac_Ip_StatusType status;
 	int err = 0;
 
 	irq_enable(cfg->rx_irq);
```
Sure.. this is a bug changed it..

sumitbatra-nxp commented on dts/bindings/ethernet/nxp,s32-gmac.yaml:
```diff
@@ -14,5 +14,8 @@ properties:
   interrupts:
     required: true
 
+  phy-handle:
```
Removed

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -167,12 +187,15 @@ static int eth_nxp_s32_init(const struct device *dev)
 	irq_enable(cfg->rx_irq);
 	irq_enable(cfg->tx_irq);
 
+	net_eth_carrier_on(ctx->iface);
```
The problem here is that, eth_nxp_s32_start gets triggered from the upper network layers net iface up <intf id> and PHY and L2 layers are not aware of what happened there.
Also I need to do net_eth_carrier_off in eth_nxp_s32_stop as if I dont do it and some network application tries to transfer packets in that state then - eth_nxp_s32_tx causes memory leak and ring buffers get stuck in used state and the network interface is not sable after that

sumitbatra-nxp replied:
What I can do is that check the network interface status before calling carrier on here

sumitbatra-nxp commented:
> Why not implement the TJA1103 phy interrupt status? That was the reason I didn't implement the link changes because I didn't want to spend a kernel worker to actively poll the phy status.

@PetervdPerk-NXP .. Done !!

sumitbatra-nxp commented:
@PetervdPerk-NXP @manuargue .. kindly re-review

dleach02 commented:
@sumitbatra-nxp please address the CI issues

sumitbatra-nxp commented:
@nashif ... Your commit - afc319e3fa8547607a5715dda57bba073ce72cf7 is causing build failures in file - subsys/testsuite/ztest/include/zephyr/ztest_test.h Line 550 and 551

zephyrproject/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_test.h:551:9: error: expected expression before 'int'
  551 |         int suite_iter, int case_iter)

zephyrproject/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_test.h:550:37: error: too few arguments to function 'z_ztest_run_test_suite'
  550 | #define ztest_run_test_suite(suite) z_ztest_run_test_suite(STRINGIFY(suite), \
      |                                     ^~~~~~~~~~~~~~~~~~~~~~

These errors are blocking this PR

manuargue reviewed (changes requested):
@sumitbatra-nxp pls rebase because current commit is not buildable.

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -78,6 +80,67 @@ static inline struct net_if *get_iface(struct eth_nxp_s32_data *ctx, uint16_t vl
 #endif
 }
 
+static void convert_phy_to_mac_state(Gmac_Ip_ConfigType *gmac_cfg, enum phy_link_speed phy_speed)
```
I considered struct phy_link_state (include/zephyr/net/phy.h) as the target for conversion (though it is phy_link_speed to be precise).
So dropping phy_to_mac_state, shall it be - phy_to_mac_config or phy_to_mac_speed ?

sumitbatra-nxp replied:
done !!

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -181,11 +257,21 @@ static int eth_nxp_s32_start(const struct device *dev)
+	irq_disable(cfg->rx_irq);
+	irq_disable(cfg->tx_irq);
+
+	if (cfg->phy_dev) {
```
if (cfg->phy_dev) is not polling the link state but rather precisely to check the case when there is no PHY

sumitbatra-nxp replied:
Lets say we already have link down from the link partner, and we do -
net ifcae down 1
This is just to avoid that case

sumitbatra-nxp replied:
You had 2 comments in one, I answered the No PHY driver comment only in my first comment :)

sumitbatra-nxp replied:
done !!

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -167,12 +230,25 @@ static int eth_nxp_s32_init(const struct device *dev)
 	irq_enable(cfg->rx_irq);
 	irq_enable(cfg->tx_irq);
 
+	if (cfg->phy_dev) {
```
Again this is exactly to handle that case.. please refer to the comment for the start function

sumitbatra-nxp replied:
done !!

sumitbatra-nxp commented on drivers/ethernet/phy/phy_tja1103.c:
```diff
@@ -128,23 +150,177 @@ static int phy_tja1103_id(const struct device *dev, uint32_t *phy_id)
+	}
+
+	/* Handling Link related Functional IRQs */
+	if (irq & (TJA1103_PHY_FUNC_IRQ_LINK_EVENT | TJA1103_PHY_FUNC_IRQ_LINK_AVAIL)) {
```
No, this will get triggered in both up and down

sumitbatra-nxp commented on drivers/ethernet/phy/phy_tja1103.c:
```diff
@@ -231,17 +428,35 @@ static const struct ethphy_driver_api phy_tja1103_api = {
+		.sem = Z_SEM_INITIALIZER(phy_tja1103_data_##n.sem, 1, 1),                       \
+	};                                                                                      \
+												\
+	static void phy_tja1103_##n##_cfg_irq_poll(void)					\
```
The only reason is that I wanted phy_tja1103_config's  **.has_irq** field to be set properly, before this function is called.

sumitbatra-nxp replied:
done !!

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -78,6 +80,67 @@ static inline struct net_if *get_iface(struct eth_nxp_s32_data *ctx, uint16_t vl
+		cfg->base->MAC_CONFIGURATION |= GMAC_MAC_CONFIGURATION_DM(gmac_cfg.Duplex);
+
+		LOG_DBG("Link up");
+		net_eth_carrier_on(ctx->iface);
```
**I have tested thoroughly.. Link never goes up if the iface is stopped**

sumitbatra-nxp replied:
done !!

sumitbatra-nxp replied:
Done !!

sumitbatra-nxp commented on drivers/ethernet/phy/phy_tja1103.c:
```diff
@@ -40,15 +41,36 @@ LOG_MODULE_REGISTER(phy_tja1103, CONFIG_PHY_LOG_LEVEL);
+#define TJA1103_PHY_FUNC_IRQ_LINK_AVAIL_EN   BIT(2)
+/* Always accessible reg for NMIs */
+#define TJA1103_ALWAYS_ACCESSIBLE	     (0x801F)
+#define TJA1103_FUSA_PASS_IRQ		     BIT(4)
```
done !!

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -167,12 +230,25 @@ static int eth_nxp_s32_init(const struct device *dev)
 	irq_enable(cfg->tx_irq);
 
+	if (cfg->phy_dev) {
+		phy_get_link_state(cfg->phy_dev, (struct phy_link_state *)&state);
```
done !!

sumitbatra-nxp commented on drivers/ethernet/eth_nxp_s32_gmac.c:
```diff
@@ -78,6 +80,67 @@ static inline struct net_if *get_iface(struct eth_nxp_s32_data *ctx, uint16_t vl
+		gmac_cfg->Speed = GMAC_SPEED_1G;
+		gmac_cfg->Duplex = GMAC_FULL_DUPLEX;
+		break;
+	default:
```
done !!

sumitbatra-nxp commented on drivers/ethernet/phy/phy_tja1103.c:
```diff
@@ -231,17 +428,35 @@ static const struct ethphy_driver_api phy_tja1103_api = {
+#define PHY_TJA1103_CFG_NO_IRQ(n)								\
+	.has_irq = 0,
+
+#define TJA1103_INITIALIZE(n)									\
```
Used https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format

PetervdPerk-NXP commented on drivers/ethernet/phy/phy_tja1103.c:
```diff
@@ -128,23 +148,172 @@ static int phy_tja1103_id(const struct device *dev, uint32_t *phy_id)
+	k_work_reschedule(&data->monitor_work, K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+}
+
+static void phy_tja1103_handle_irq(const struct device *port, struct gpio_callback *cb,
```
MDIO is a slow blocking protocol. You don't want to do these transfers in irq context since it destroys determinism of the system.

sumitbatra-nxp replied:
Every ISR needs to atleast check the reason of interrupt and try to ack it by accessing the registers of the device that caused it.
Its just that in Ethernet PHY case these device registers are not memory mapped so what's the alternative then ?
Also check out TJA1103 Linux driver, it does the same-
drivers/net/phy/nxp-c45-tja11xx.c (nxp_c45_handle_interrupt)

sumitbatra-nxp replied:
I have done it using an offload thread and a semaphore..  any comments are welcome

sumitbatra-nxp committed:
drivers: eth: s32 gmac: Handle link change

These changes enable -
1. S32 gmac driver to gracefully handle
net iface down and up net shell commands and
2. Link connect or disconnect events between
Ethernet PHY and its link partner.

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp committed:
drivers: eth: phy: tja1103: Handle link change

drivers: eth: phy: tja1103: Handle link change
These changes enable -
TJA1103 driver to gracefully handle Link connect or disconnect events
between Ethernet PHY and its link partner and notify it to the
upper network layers

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>
